### PR TITLE
Fixes #2491 Correctly delete all files for dates archives

### LINKED ISSUE #2491: .no-webp files not deleted from date archives when clearing related cache
**Before submitting an issue please check that you’ve completed the following steps:**
- Made sure you’re on the latest version ✔
- Used the search feature to ensure that the bug hasn’t been reported before ✔

**Describe the bug**
The `.no-webp` file created during webp caching isn't cleared from the date archives when a post is updated. 

**To Reproduce**
1. Enable webp caching in WP Rocket. Make sure `.no-webp` files are generated for pages without webp images. 
2. Disable Preload. 
3. Visit the date archives to generate it's cache. 
4. Update the post. It's cache and all related caches will be cleared. 
5. However, `.no-webp` files will not be cleared. 

**Expected behavior**
The `.no-webp` file should be deleted as well. In case the archive page now has webp images, having the `.no-webp` file here will cause issue. 

**Screenshots**
Video: https://youtu.be/D8cA_h06WoA

**Backlog Grooming (for WP Media dev team use only)**
- [x] Reproduce the problem
- [x] Identify the root cause
- [x] Scope a solution
- [x] Estimate the effort

### PR DESCRIPTION
**Pull request by remyperona**
Fixes #2491 and more, as it's also going to correctly delete all the non-default cache files, like the mobile one, or the one used for dynamic cookies.

- [x] Deprecate `get_rocket_post_dates_urls()`
- [x] Remove dates archives from default purge urls list in `rocket_get_purge_urls()`
- [x] Add `WP_Rocket\Engine\Cache\Purge` class
- [x] Add callback method to purge the dates archives in `PurgeActionsSubscriber`
- [x] Add tests

### CODE REVIEW

**Comment by Tabrisrp** [comment]
I'm thinking we will want to refactor part of the code happening in `purge_dates_archives()`, as it's very close to what Ahmed wrote in this PR for `rocket_clean_home()` https://github.com/wp-media/wp-rocket/pull/2688

Possibly have a function to deal with files + pagination entries deletion only.

**Review by hellofromtonya — CHANGES_REQUESTED** [review]
The `Purge::purge_dates_archive()` has too many different tasks in it. Break it up into separate, specific increments of behavior via private methods.

File: inc/Engine/Cache/Purge.php [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+	public function purge_dates_archives( $post ) {
+		global $wp_rewrite;
+
+		foreach ( $this->get_dates_archives( $post ) as $url ) {
+			$parsed_url = get_rocket_parse_url( $url );
+
+			/** This filter is documented in inc/front/htaccess.php */
+			if ( apply_filters( 'rocket_url_no_dots', false ) ) {
+				$parsed_url['host'] = str_replace( '.', '_', $parsed_url['host'] );
+			}
```
[the remark this answers is not in the document]
Comment (Tabrisrp) [reply inc/Engine/Cache/Purge.php]: Done in https://github.com/wp-media/wp-rocket/pull/2736/commits/e6325eee259298f3e7329236f197c7d58cbaae04

File: inc/Engine/Cache/Purge.php [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+					if ( $item->isFile() ) {
+						$this->filesystem->delete( $item->getPathname() );
+					}
+				}
+
+				$pagination_dir = $path . DIRECTORY_SEPARATOR . $wp_rewrite->pagination_base;
+
+				if ( $this->filesystem->is_dir( $pagination_dir ) ) {
+					rocket_rrmdir( $pagination_dir, [], $this->filesystem );
+				}
```
[the remark this answers is not in the document]
Comment (Tabrisrp) [reply inc/Engine/Cache/Purge.php]: Done in https://github.com/wp-media/wp-rocket/pull/2736/commits/e6325eee259298f3e7329236f197c7d58cbaae04

File: inc/Engine/Cache/Purge.php [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+				try {
+					$iterator = new DirectoryIterator( $path );
+				} catch ( Exception $e ) {
+					// No action required, as logging not enabled.
+					$iterator = [];
+				}
+
+				foreach ( $iterator as $item ) {
```
[the remark this answers is not in the document]
Comment (Tabrisrp) [reply inc/Engine/Cache/Purge.php]: Done in https://github.com/wp-media/wp-rocket/pull/2736/commits/e6325eee259298f3e7329236f197c7d58cbaae04

**Review by hellofromtonya — APPROVED** [review]

**Comment by Tabrisrp** [comment]
@wp-media/dev-qa Ready for QA.

With this change, when a post is updated, the cache files & hidden files for dates archives should be correctly deleted, whatever format they use (index.html, index-mobile.html_gzip, index-https-webp.html, etc). Hidden files are `.mobile-active` and `.no-webp`

**Review by crystinutzaa — APPROVED** [review]

**Review by arunbasillal — APPROVED** [review]
- Recreated the issue with current master branch. Left over files
image
- Repeated the same in fix branch. Everything was cleared 👍 
- Cleared WP Rocket cache, everything was cleared 👍 
- Executed `rocket_clean_post` and it worked as before 👍 
- Nothing logged in `debug.log` 👍 

@Tabrisrp Do you think I missed anything? Should I test any more scenarios?

**Comment by Tabrisrp** [comment]
The method is hooked on `rocket_clean_post`, so you pretty much did what was needed 👍

**Pull request merged**